src/cache/core/cache.ts

```diff
@@ -141,6 +174,46 @@ export abstract class ApolloCache<TSerialized> implements DataProxy {
     });
   }
 
+  public watchFragment<TData = any, TVars = OperationVariables>(
+    options: WatchFragmentOptions<TData, TVars>
+  ): Observable<WatchFragmentResult<TData>> {
+    const { fragment, fragmentName, from, optimistic = true } = options;
+
+    const diffOptions: Cache.DiffOptions<TData, TVars> = {
+      returnPartialData: true,
+      id: typeof from === "string" ? from : this.identify(from),
+      query: this.getFragmentDoc(fragment, fragmentName),
+      optimistic,
+    };
+
+    let latestDiff = this.diff<TData>(diffOptions);
+
+    return new Observable((observer) => {
+      return this.watch<TData, TVars>({
+        ...diffOptions,
+        immediate: true, // is this necessary? tests pass without it
+        query: this.getFragmentDoc(fragment, fragmentName),
+        callback(diff) {
+          const result = {
+            data: diff.result as DeepPartial<TData>,
+            complete: !!diff.complete,
+          } as WatchFragmentResult<TData>;
+
+          if (diff.missing) {
+            result.missing = mergeDeepArray(
+              diff.missing.map((error) => error.missing)
+            );
+          }
+
+          if (!equal(diff, latestDiff)) {
+            latestDiff = diff;
+            observer.next(result);
+          }
+        },
```

Comment:
[nit] You might consider moving this `if` to the top of the `callback` function and returning early if these values are equal. No need to create a `result` object if we aren't going to emit it.

```suggestion
        callback(diff) {
          if (equal(diff, latestDiff)) {
            return;
          }
          
          const result = {
            data: diff.result as DeepPartial<TData>,
            complete: !!diff.complete,
          } as WatchFragmentResult<TData>;

          if (diff.missing) {
            result.missing = mergeDeepArray(
              diff.missing.map((error) => error.missing)
            );
          }
          
          latestDiff = diff;
          observer.next(result);
        },
```

Reply:
Updated in https://github.com/apollographql/apollo-client/pull/11465/commits/ea48ea3c6dea5dfce34d50e0e8505c0e8badfa09